### PR TITLE
difftree for git.Trees

### LINKED ISSUE #82: increase performance of DiffTree, applying the same algorithm being use at git

### PR DESCRIPTION
**Pull request by alcortesm**
Last PR to fix https://github.com/src-d/go-git/issues/82:

This PR modifies the difftree package itself.  The old version extracted the files in both trees and compare them by hand.  The new version turn the trees into `merkletrie.Noders` and call the `merkletrie.Difftree` function on them.

How to review this PR:
- treenoder.go: defines the `treeNoder` type that wraps a `git.Tree` and implements `merkletrie.Noder`.
- change.go: defines the type of the output of a difftree operation. The type is the same as before, but I have moved it into its own file to keep the package organized.  The old package defines the `Action` type too (insert, delete, modify), now, we reuse `merkletrie.Action` and it is no longer a field, but a method.
- change_adaptor.go: defines functions to turn `merkletrie.Change`s into `difftree.Change`s.
- difftree.go: before this patch this file holds all the logic to do a difftree, now it just turns the `git.Trees` into `treeNoders`, call `merkletrie.difftree` on them, and turns the resulting `merkletrie.Change`s into `difftree.Change`s.

The only interesting piece of code here is that noders don't have the concept of `mode` (file permissions). The treenoder type codifies `git.Tree` modes into the `merkletrie.Noder` hash, so changes in the mode of a file are detected as modifications, just as the original `git diff-tree` command does.

### CODE REVIEW

File: plumbing/difftree/change_adaptor.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package difftree
+
+// The folowing functions transforms changes types form the merkletrie
```
Comment (smola) [inline plumbing/difftree/change_adaptor.go]: `transforms -> transform`

File: plumbing/difftree/change_adaptor.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// package to changes types from this package.
+//
+// newChange      transforms a merkletrie.Change   into a Change.
```
Comment (smola) [inline plumbing/difftree/change_adaptor.go]: these comments look quite redundant, it's just their signature, right?

File: plumbing/difftree/change_adaptor_test.go [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+	"sort"
+
+	"srcd.works/go-git.v4/plumbing"
+	"srcd.works/go-git.v4/plumbing/object"
+	"srcd.works/go-git.v4/plumbing/storer"
+	"srcd.works/go-git.v4/storage/filesystem"
+	"srcd.works/go-git.v4/utils/merkletrie"
+	"srcd.works/go-git.v4/utils/merkletrie/noder"
+
+	fixtures "github.com/src-d/go-git-fixtures"
```
Comment (smola) [inline plumbing/difftree/change_adaptor_test.go]: isn't this redundant? the package is already named fixtures

File: plumbing/difftree/change_adaptor.go (hunk 1 above)
Comment (alcortesm) [inline plumbing/difftree/change_adaptor.go]: fixed.

File: plumbing/difftree/change_adaptor.go (hunk 2 above)
Comment (alcortesm) [inline plumbing/difftree/change_adaptor.go]: Yes, it is just their signature, and it is totally redundant. Removed.

File: plumbing/difftree/change_adaptor_test.go (hunk 3 above)
Comment (alcortesm) [inline plumbing/difftree/change_adaptor_test.go]: Yes, it is redundant leftover. Removed.

File: plumbing/difftree/change.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if c.From != empty {
+		return c.From.Name
+	}
+
+	return c.To.Name
+}
+
+// ChangeEntry values represent a node that has suffered a change.
+type ChangeEntry struct {
+	Name      string           // full path of the node using "/" as separator
```
Comment (mcuadros) [inline plumbing/difftree/change.go]: public struct fields should be documented on top of the definiton

File: plumbing/difftree/change.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	Tree      *object.Tree     // parent tree of the node that has changed
+	TreeEntry object.TreeEntry // the entry of the node
+}
+
+// Changes represents a collection of changes between two git trees.
```
Comment (mcuadros) [inline plumbing/difftree/change.go]: Maybe add to the documentation that implements the sort interface will be nice

File: plumbing/difftree/change.go (hunk 4 above)
Comment (alcortesm) [inline plumbing/difftree/change.go]: Yeah, old habits die hard. Fixed.

File: plumbing/difftree/change.go (hunk 5 above)
Comment (alcortesm) [inline plumbing/difftree/change.go]: Yes, fixed!.

Comment (alcortesm) [inline plumbing/difftree/change.go]: rebased to master.

**Pull request merged**